internal/tfr/getter.go

```diff
@@ -0,0 +1,259 @@
+		var opts []getter.ClientOption
+		if tfrGetter.client != nil {
+			opts = tfrGetter.client.Options
+		}
+		return getter.Get(dstPath, source, opts...)
+	}
+
+	// We have a subdir, time to jump some hoops
+	return tfrGetter.getSubdir(ctx, dstPath, source, path.Join(subDir, moduleSubDir))
+}
+
+// GetFile is not implemented for the Terraform module registry Getter since the terraform module registry doesn't serve
+// a single file.
+func (tfrGetter *TerraformRegistryGetter) GetFile(dst string, src *url.URL) error {
+	return errors.WithStackTrace(fmt.Errorf("GetFile is not implemented for the Terraform Registry Getter"))
+}
+
+// getSubdir downloads the source into the destination, but with the proper subdir.
+func (tfrGetter *TerraformRegistryGetter) getSubdir(ctx context.Context, dstPath, sourceURL, subDir string) error {
+	// Create a temporary directory to store the full source. This has to be a non-existent directory.
+	tempdirPath, tempdirCloser, err := safetemp.Dir("", "getter")
+	if err != nil {
+		return err
+	}
+	defer tempdirCloser.Close()
+
+	var opts []getter.ClientOption
+	if tfrGetter.client != nil {
+		opts = tfrGetter.client.Options
+	}
+	// Download that into the given directory
+	if err := getter.Get(tempdirPath, sourceURL, opts...); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// Process any globbing
+	sourcePath, err := getter.SubdirGlob(tempdirPath, subDir)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// Make sure the subdir path actually exists
+	if _, err := os.Stat(sourcePath); err != nil {
+		details := fmt.Sprintf("could not stat download path %s (error: %s)", sourcePath, err)
+		return errors.WithStackTrace(ModuleDownloadErr{sourceURL: sourceURL, details: details})
+	}
+
+	// Copy the subdirectory into our actual destination.
+	if err := os.RemoveAll(dstPath); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// Make the final destination
+	if err := os.MkdirAll(dstPath, 0755); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// We use a temporary manifest file here that is deleted at the end of this routine since we don't intend to come
+	// back to it.
+	manifestFname := ".tgmanifest"
+	manifestPath := filepath.Join(dstPath, manifestFname)
+	defer os.Remove(manifestPath)
+	return util.CopyFolderContentsWithFilter(sourcePath, dstPath, manifestFname, func(path string) bool { return true })
+}
+
+// getModuleRegistryURLBasePath uses the service discovery protocol
+// (https://www.terraform.io/docs/internals/remote-service-discovery.html)
+// to figure out where the modules are stored. This will return the base
+// path where the modules can be accessed
+func getModuleRegistryURLBasePath(ctx context.Context, domain string) (string, error) {
+	sdURL := url.URL{
+		Scheme: "https",
+		Host:   domain,
+		Path:   serviceDiscoveryPath,
+	}
+	bodyData, _, err := httpGETAndGetResponse(ctx, sdURL)
+	if err != nil {
+		return "", err
+	}
+
+	var respJSON map[string]interface{}
```

Comment:
Any reason not to parse into a more well-defined, typed struct?

Reply:
Ah originally I had the misconception that golang didn't support partial json parsing, but it looks like it does, so went ahead and created a struct: 
b51591c